src/holon/rule_engine/repositories/repository_base.py

```diff
@@ -111,43 +105,46 @@ def get_subset_range(
 
         return self.__class__(objects)
 
-    # TODO ERIK
-    def get(self, id: int) -> object:
-        """Get an item in the objects list by id"""
+    def get(self, id: int) -> PolymorphicModel:
+        """Get an item in the objects list by id. Return None if object was not found"""
 
-        raise NotImplementedError()
+        return next((object for object in self.objects if object.id == id), None)
 
-    def all(self) -> list[object]:
+    def all(self) -> list[PolymorphicModel]:
         """Return all objects in the repository"""
+
         return self.objects
 
     def len(self) -> int:
         """Return the number of objects in the repository"""
+
         return len(self.objects)
 
     def update_attribute(self, id: int, attribute_name: str, value):
-        """
-        Select an object by id in the repository and set it's attribute attribute_name to value
-        <CHANGES INTERNAL STATE>
-        """
+        """Select an object by id in the repository and set it's attribute attribute_name to value"""
 
-        raise NotImplementedError()
+        object_index = next((i for i, object in enumerate(self.objects) if object.id == id), -1)
 
-    # TODO ERIK
-    def add(self, object):
-        """
-        Add an object to the repository
-        <CHANGES INTERNAL STATE>
-        """
-        raise NotImplementedError()
+        if object_index < 0:
+            raise IndexError(f"{self.base_model_type.__name__} object with id {id} not found")
 
-    # TODO ERIK
-    def remove(self, object):
-        """
-        Remove an item from the repository
-        <CHANGES INTERNAL STATE>
-        """
-        raise NotImplementedError()
+        setattr(self.objects[object_index], attribute_name, value)
+
+    def add(self, object: PolymorphicModel):
+        """Add an object to the repository"""
+
+        # check if object base type is correct
+        if not self.base_model_type.__name__ == utils.get_base_polymorphic_model(object).__name__:
+            raise ValueError(
+                f"Can only insert objects of type {self.base_model_type.__name__}. Object type for attempted insertion: {utils.get_base_polymorphic_model(object).__name__}"
+            )
+
+        self.objects.append(object)
```

Comment:
id's for added object need to be unique/changed

Reply:
The caller would probably like to know the id. So either return the object here, or let the caller set the id beforehand.

In either case we can have a global atomic counter which can create new id's.

I'm in favor of always using UUID's and never integer ids so id creation doesn't have to be coordinated. But that would be a different change.

Reply:
I made an id counter that starts at the max id of the provided objects + 1 and increments for every add action.